### ISSUE TITLE
Search Space for multi-trial NAS

### ISSUE THREAD

**Issue by NotSure2732** [microsoft/nni#4731]
**Describe the issue**:
I am currently trying to get familiar with how to define a search space for NAS in nni. Are there any more in-depth examples of defined search spaces other than in the example notebook for multi-trial NAS? Because for example they do not show how to properly vary the size of the network, only how to vary filter_size and the use of skip_connections, I am interested in the use of  nn.Repeat or nn.ValueChoice. 
More specifically regarding the implementation of a search space for CNNs:
I have successfully defined models that vary in network width (via ValueChoice), but I am struggling to vary network depth. The only way I managed to do this is by using nn.Repeat, making a block that repeats (min,max) times and keep in_channels=out_channels. For a proper search space I would like to define a model that can have a different number of layers and for each layer also vary width of the network. But I do not know how to combine them or use nn.Repeat in a way that allows in_channels!=out_channels.

Any help is highly appreciated!

**Environment**:
- NNI version:
- Training service (local|remote|pai|aml|etc):
- Client OS:
- Server OS (for remote mode only):
- Python version:
- PyTorch/TensorFlow version:
- Is conda/virtualenv/venv used?:
- Is running in Docker?:

**Configuration**:
 - Experiment config (remember to remove secrets!):
 - Search space:

**Log message**:
 - nnimanager.log:
 - dispatcher.log:
 - nnictl stdout and stderr:
 

**How to reproduce it?**:

**Comment by matluster** [comment]
`Repeat` supports a lambda expression to instantiate different blocks for different indices. Maybe that's what you need?

Not sure whether it's included in v2.6 release though.

**Comment by NotSure2732** [comment]
That would certainly be helpful I think, are there any use examples?
Or if you don't mind, could you share a simple example.
Thanks you!

**Comment by NotSure2732** [comment]
A few follow-up questions, which belong to the same category of questions:
1. Is there a possibility to use the chosen values in ValueChoice or even in repeat in the following code. E.g. one issue would be the first input size of a fully connected layer after the last convolutional layer varied the output size (width). It is not possible to simply use a variable for the ValueChoice, since in the fully-connected layer that value needs to be multiplied by the size of the image, which is not possible using a ValueChoice variable (yet).
2. If I use the mentioned way of using nn.Repeat (using different blocks for different indices) how would I make sure that after the input for layers after the chosen block fits to the output of that block. Would they all need to end in the same size, or is there a way to make different possible layers depending on which block/ layer has been chosen before?
3. Using nn.Repeat using 1 layer and a tuple to repeat it, is there a way to get the activation function after every repeat instead of after all repeated layers?

Thank you for help in advance,
I really appreciate it!

**Comment by matluster** [comment]
In v2.7 documentation: https://nni.readthedocs.io/en/v2.7/reference/nas/search_space.html#nni.retiarii.nn.pytorch.Repeat

**Comment by NotSure2732** [comment]
I tried to instantiate Blocks and use them in LayerChoice, but immediately get the following error:
     [2022-04-14 19:08:27] INFO (nni.experiment/MainThread) Creating experiment, Experiment ID: 56bn7iwc
[2022-04-14 19:08:27] INFO (nni.experiment/MainThread) Connecting IPC pipe...
[2022-04-14 19:08:27] INFO (nni.experiment/MainThread) Starting web server...
[2022-04-14 19:08:28] INFO (nni.experiment/MainThread) Setting up...
[2022-04-14 19:08:28] INFO (nni.runtime.msg_dispatcher_base/Thread-6) Dispatcher started
[2022-04-14 19:08:28] INFO (nni.retiarii.experiment.pytorch/MainThread) Web UI URLs: http://127.0.0.1:8745/ http://137.56.7.110:8745/
[2022-04-14 19:08:28] INFO (nni.retiarii.experiment.pytorch/MainThread) Start strategy...
[2022-04-14 19:08:28] INFO (root/MainThread) Successfully update searchSpace.
[2022-04-14 19:08:28] INFO (nni.retiarii.strategy.bruteforce/MainThread) Random search running in fixed size mode. Dedup: on.
[2022-04-14 19:08:28] INFO (nni.retiarii.experiment.pytorch/MainThread) Strategy exit
[2022-04-14 19:08:28] INFO (nni.retiarii.experiment.pytorch/MainThread) Waiting for experiment to become DONE (you can ctrl+c if there is no running trial jobs)...

This is how I tried to use simple Blocks:

from nni.retiarii.serializer import model_wrapper
import torch.nn.functional as F
import nni.retiarii.nn.pytorch as nn
      
class Block1(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(128, 256, 3, stride=1,padding=1)
        self.conv2 = nn.Conv2d(256, 512, 3, stride=1, padding=1)
        
    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        return x
    
class Block2(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(128, 512, 3, stride=1,padding=1)
        
    def forward(self, x):
        x = F.relu(self.conv1(x))
        return x
    
@model_wrapper
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 64, 3, stride=1,padding=1)
        self.conv2 = nn.Conv2d(64, 128, 3, stride=1, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv3 = nn.LayerChoice([Block1(),Block2()])
        self.conv4 = nn.Conv2d(512,1024 , 3, stride=1, padding=1)
        self.fc1 = nn.Linear(1024*8*8, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)
        
    def forward(self, x):

        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = self.pool(x)
        x = self.conv3(x)  
        x = self.pool(x)
        x = F.relu(self.conv4(x))    
        x = x.reshape(x.shape[0],-1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

model = Net()

If someone is able to tell me why this does not work/ how to properly do this I would be very grateful!

**Comment by matluster** [comment]
v2.7 is not released yet. It will be available next week. To try it now, you need to [install from source](https://nni.readthedocs.io/en/latest/notes/build_from_source.html) (trust me, it's not as scary as it looks).

> How would one change this in order to make the ReLu appear after every conv3?

You can do this:

```python
self.repeat = nn.Repeat(nn.Sequential(Conv2d(...), nn.ReLU()), (1, 3))
```

> I tried to instantiate Blocks and use them in LayerChoice, but immediately get the following error:

Where is your error message?

**Comment by NotSure2732** [comment]
Ah ok, thank you for the info and the solution!

Sorry, not exactly an error message;
[2022-04-14 19:08:28] INFO (nni.retiarii.experiment.pytorch/MainThread) Strategy exit
this shows up immediately after starting the experiment

**Comment by NotSure2732** [comment]
I think I figured it out. Since I ran the experiment with only one mutation (so two different architectures) and used a concurrency of 2 trials, the strategy was immediately able to finish. So now I added another choice (resulting in 4 models) and it worked as expected. So it was my misjudgement, not an error.

**Comment by NotSure2732** [comment]
Regarding the use of v2.7:
I tried to install from source but it still shows nni version 2.6.1:

!git clone https://github.com/microsoft/nni.git
Cloning into 'nni'...
remote: Enumerating objects: 51807, done.
remote: Counting objects: 100% (2442/2442), done.
remote: Compressing objects: 100% (1294/1294), done.
remote: Total 51807 (delta 1311), reused 1917 (delta 1089), pack-reused 49365
Receiving objects: 100% (51807/51807), 60.97 MiB | 4.55 MiB/s, done.
Resolving deltas: 100% (30024/30024), done.

!pip show nni
Name: nni
Version: 2.6.1
Summary: Neural Network Intelligence project
Home-page: https://github.com/Microsoft/nni
Author: Microsoft NNI Team
Author-email: nni@microsoft.com
License: MIT
Location: /Users/opt/anaconda3/lib/python3.7/site-packages
Requires: astor, cloudpickle, colorama, filelock, hyperopt, json-tricks, numpy, pandas, prettytable, psutil, PythonWebHDFS, pyyaml, requests, responses, schema, scikit-learn, scipy, typeguard, typing-extensions, websockets
Required-by: 


I also tested it by using a ValueChoice object arithmetically, but get an error. Showing it does not have the functionality of v2.7.
Did I do something wrong? I removed the nni directory since it first showed an existing path, but still the same outcome.

**Comment by ultmaster** [comment]
Please check the build from source documentation as I said before. You need to run something like python setup.py develop.
